### PR TITLE
refactor: Replace g_genesis_wait_cv with m_tip_block_cv

### PR DESCRIPTION
**Pull request by maflcko**
`g_genesis_wait_cv` is similar to `m_tip_block_cv` but shuffling everything through a redundant `boost::signals2`.

So remove it, along with some other dead code, as well as minor fixups.

### CODE REVIEW

**Comment by TheCharlatan** [comment]
Nice, Concept ACK

File: src/init.cpp [hunk 1]
```diff
@@ -1822,17 +1798,17 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
 
     // ********************************************************* Step 12: start node
 
-    //// debug print
```
Comment (Sjors) [inline src/init.cpp:1825]: fa964f82c52a7d0fd40c5133cf5fc18ef13819e5 nit: seems everything below here could be its own commit

**Review by Sjors — APPROVED** [review]
utACK fa964f82c52a7d0fd40c5133cf5fc18ef13819e5

Dropping `g_genesis_wait_cv` is a nice simplification.

File: src/node/interfaces.cpp [hunk 2]
```diff
@@ -945,7 +945,7 @@ class MinerImpl : public Mining
         const auto deadline{now + timeout};
         {
             WAIT_LOCK(notifications().m_tip_block_mutex, lock);
-            while ((notifications().m_tip_block == uint256() || notifications().m_tip_block == current_tip) && !chainman().m_interrupt) {
+            while ((notifications().m_tip_block == uint256::ZERO || notifications().m_tip_block == current_tip) && !chainman().m_interrupt) {
```
Comment (theuni) [inline src/node/interfaces.cpp:948]: While you're at it, it'd be more idiomatic to use the `wait_for` that takes a predicate, rather than wrapping a `wait_until` in a `while()`.

**Review by theuni — COMMENTED** [review]
Nice improvements, lgtm.

File: src/node/interfaces.cpp (hunk 2 above)
Comment (ryanofsky) [reply src/node/interfaces.cpp:948]: re: https://github.com/bitcoin/bitcoin/pull/30967#discussion_r1775211951

I'm not sure about the steady clock thing, but I think the reason this code is written to use a while loop is because it wants to wake up every 1000ms to check m_interrupt in case `m_interrupt` was signaled without `m_tip_block_cv` being signalled. Since I posted another change above to ensure that signalling `m_interrupt` will reliably signal `m_tip_block_cv`, I think this code could be changed to use a simple wait predicate and drop the 1000ms polling.

File: src/rpc/server.h [hunk 3]
```diff
@@ -9,7 +9,6 @@
 #include <rpc/request.h>
 #include <rpc/util.h>
 
-#include <any>
```
Comment (TheCharlatan) [inline src/rpc/server.h:12]: Nit: Could forward-declare `NodeContext` here instead.

**Review by ryanofsky — APPROVED** [review]
Code review ACK fa61644768c3f6932860d1c5e2c6c0a38d5257e5.

Overall very nice changes, but I think the first commit fa5ac85d19a387c2e12fc0833368e1240a3dcf51 has a drawback because it adds a `NodeContext` argument to `RPCStop` when ideally RPC code should be agnostic to node initialization and not depend on node types. Would suggest following change to clean it up, which could be squashed into the first commit:

- ef271bc75e7bea8c82ce8667fb22e9b25dca5712 refactor: Split up NodeContext shutdown_signal and shutdown_request

This would also enable additional simplification in AppInitMain which could be squashed into commit faaf888eb275cecc2d2803c5756afac633e0e2a4:

- 35d0986f43264da69384208701d659990f50dbd6 refactor: Simplify AppInitMain genesis block wait loop

(Suggested commits are from branch https://github.com/ryanofsky/bitcoin/tree/review.30967.2-edit based on this PR)

**Review by sedited — APPROVED** [review]
ACK fa61644768c3f6932860d1c5e2c6c0a38d5257e5

File: src/node/interfaces.cpp (hunk 2 above)
Comment (ryanofsky) [reply src/node/interfaces.cpp:949]: re: https://github.com/bitcoin/bitcoin/pull/30967#discussion_r1775163810

I think Cory's suggestion could be implemented here with:

```diff
--- a/src/node/interfaces.cpp
+++ b/src/node/interfaces.cpp
@@ -940,17 +940,11 @@ public:
 
     BlockRef waitTipChanged(uint256 current_tip, MillisecondsDouble timeout) override
     {
-        // Interrupt check interval
-        const MillisecondsDouble tick{1000};
-        auto now{std::chrono::steady_clock::now()};
-        const auto deadline{now + timeout};
         {
             WAIT_LOCK(notifications().m_tip_block_mutex, lock);
-            while ((notifications().m_tip_block == uint256::ZERO || notifications().m_tip_block == current_tip) && !chainman().m_interrupt) {
-                now = std::chrono::steady_clock::now();
-                if (now >= deadline) break;
-                notifications().m_tip_block_cv.wait_until(lock, std::min(deadline, now + tick));
-            }
+            notifications().m_tip_block_cv.wait_for(lock, timeout, [&] {
+                return (notifications().m_tip_block != current_tip && notifications().m_tip_block != uint256::ZERO) || chainman().m_interrupt;
+            });
         }
         // Must release m_tip_block_mutex before locking cs_main, to avoid deadlocks.
         LOCK(::cs_main);
```

**Review by ryanofsky — APPROVED** [review]
Code review ACK fa7c2c9f18f5468baf81c04177b75e2b684a327a

File: src/rpc/server.cpp [hunk 4]
```diff
@@ -303,9 +303,6 @@ void StopRPC(const std::any& context)
         LogDebug(BCLog::RPC, "Stopping RPC\n");
         WITH_LOCK(g_deadline_timers_mutex, deadlineTimers.clear());
         DeleteAuthCookie();
-        node::NodeContext& node = EnsureAnyNodeContext(context);
-        // The notifications interface doesn't exist between initialization step 4a and 7.
```
Comment (Sjors) [inline src/rpc/server.cpp:307]: 86bddb439bff846630b1efdea777119439ba56b2: did you mean to drop this comment? (the ` if (node.notifications)` check is still there)

File: src/node/interfaces.cpp (hunk 2 above)
Comment (Sjors) [reply src/node/interfaces.cpp:949]: > I think the reason this code is written to use a while loop is because it wants to wake up every 1000ms to check m_interrupt in case `m_interrupt` was signaled without `m_tip_block_cv` being signalled

IIRC that's indeed why.

**Comment by Sjors** [comment]
Commit 86bddb439bff846630b1efdea777119439ba56b2 is a bit over my head, but seems like a good idea.

Lightly reviewed fa7c2c9f18f5468baf81c04177b75e2b684a327a.

**Review by ryanofsky — APPROVED** [review]
Code review ACK fa877f4c5f9abb057f1476f21a063e0f687d0ad1. Since last review, just added new commit with another simplification.

---

Could consider squashing the new commit:

-  "refactor: Use wait_for predicate to check for interrupt" (fa877f4c5f9abb057f1476f21a063e0f687d0ad1)

into earlier commit:

-  "refactor: Remove dead code in waitTipChanged" (fafffc8b5cc25adf278031561f060023a14e4387)

since they are touching same code, and there should be no need to have one commit simplifying some arithmetic followed by another commit removing the arithmetic.

Could also consider moving the doc commit:

- "doc: Clarify waitTipChanged docs" (fae9bd8dd919b55faf3e31371d3e1100f2ab704f)

to be first, so it's clear the documentation fixes in it apply to current code, not just new code.

File: src/node/interfaces.cpp (hunk 2 above)
Comment (theuni) [reply src/node/interfaces.cpp]: Smells familiar. I wonder if it's the same root issue as this: https://github.com/bitcoin/bitcoin/pull/30462#discussion_r1679972352

File: src/node/interfaces.cpp [hunk 5]
```diff
@@ -948,7 +948,7 @@ class MinerImpl : public Mining
         if (deadline < now) deadline = std::chrono::steady_clock::time_point::max();
         {
             WAIT_LOCK(notifications().m_tip_block_mutex, lock);
-            while ((notifications().m_tip_block == uint256() || notifications().m_tip_block == current_tip) && !chainman().m_interrupt) {
+            while ((notifications().m_tip_block == uint256::ZERO || notifications().m_tip_block == current_tip) && !chainman().m_interrupt) {
```
Comment (ryanofsky) [inline src/node/interfaces.cpp]: In commit "refactor: Add missing GUARDED_BY(m_tip_block_mutex)" (fa47d893acdd18132204810dd95880ab21abf415)

It seems like it would be a little better to make this change in the next commit "refactor: Use wait_for predicate to check for interrupt" (faa3f411f6842bf1201b25d6391b1913e1b72439) instead of this one so this code is not changing more than once.

File: src/node/interfaces.cpp [hunk 6]
```diff
@@ -940,19 +940,12 @@ class MinerImpl : public Mining
 
     BlockRef waitTipChanged(uint256 current_tip, MillisecondsDouble timeout) override
     {
-        // Interrupt check interval
-        const MillisecondsDouble tick{1000};
-        auto now{std::chrono::steady_clock::now()};
-        auto deadline = now + timeout;
-        // std::chrono does not check against overflow
-        if (deadline < now) deadline = std::chrono::steady_clock::time_point::max();
+        if (timeout > std::chrono::years{100}) timeout = std::chrono::years{100}; // Upper bound to avoid UB in std::chrono
```
Comment (ryanofsky) [inline src/node/interfaces.cpp:943]: In commit "refactor: Use wait_for predicate to check for interrupt" (faa3f411f6842bf1201b25d6391b1913e1b72439)

Might be better to avoid the 100 year hack by switching to wait_until

```diff
--- a/src/node/interfaces.cpp
+++ b/src/node/interfaces.cpp
@@ -940,10 +940,10 @@ public:
 
     BlockRef waitTipChanged(uint256 current_tip, MillisecondsDouble timeout) override
     {
-        if (timeout > std::chrono::years{100}) timeout = std::chrono::years{100}; // Upper bound to avoid UB in std::chrono
         {
             WAIT_LOCK(notifications().m_tip_block_mutex, lock);
-            notifications().m_tip_block_cv.wait_for(lock, timeout, [&]() EXCLUSIVE_LOCKS_REQUIRED(notifications().m_tip_block_mutex) {
+            const auto deadline{std::chrono::steady_clock::now() + timeout};
+            notifications().m_tip_block_cv.wait_until(lock, deadline, [&]() EXCLUSIVE_LOCKS_REQUIRED(notifications().m_tip_block_mutex) {
                 return (notifications().m_tip_block != current_tip && notifications().m_tip_block != uint256::ZERO) || chainman().m_interrupt;
             });
         }
```

**Review by ryanofsky — APPROVED** [review]
Code review ACK fa24d8dbddebd61b9c418bd6773c333ad72bed67. Fixed wait_for overflow bug since last review

File: src/node/interfaces.cpp [hunk 7]
```diff
@@ -940,19 +940,12 @@ class MinerImpl : public Mining
...
         {
             WAIT_LOCK(notifications().m_tip_block_mutex, lock);
-            while ((notifications().m_tip_block == uint256() || notifications().m_tip_block == current_tip) && !chainman().m_interrupt) {
-                now = std::chrono::steady_clock::now();
-                if (now >= deadline) break;
-                notifications().m_tip_block_cv.wait_until(lock, std::min(deadline, now + tick));
-            }
+            notifications().m_tip_block_cv.wait_for(lock, timeout, [&]() EXCLUSIVE_LOCKS_REQUIRED(notifications().m_tip_block_mutex) {
+                return (notifications().m_tip_block != current_tip && notifications().m_tip_block != uint256::ZERO) || chainman().m_interrupt;
```
Comment (Sjors) [inline src/node/interfaces.cpp:947]: I think that since `waitTipChanged` takes a `current_tip` argument the `m_tip_block != uint256::ZERO` check is no longer needed. Though it seems fine to keep it.

IIRC the reason I added it was in case `waitTipChanged` was called during early node initialization, before `m_tip_block` is set to the genesis block.

That scenario is now handled by `m_tip_block != current_tip`.

In term of how the Template Provider works, it used to rely on this check, but now it just waits for `getTip()` to return a value before calling `waitTipChanged()`. .

When implemented as part of Bitcoin Core, as in #29432, it can't call `waitTipChanged()` earlier, because it has no way of knowing what to set for `current_tip`.

When implemented as an external application (e.g. https://github.com/Sjors/bitcoin/pull/48) it might know the tip from some other source before `m_tip_block` is set. So it could call `waitTipChanged()` very early in node initialization, but that still doesn't require the `m_tip_block != uint256::ZERO` check.

**Comment by Sjors** [comment]
re-utACK cccc5b19fba6d04e586f59bff5f5b2cdef3dffba

The main change since my last review is in 444424b96deef72780f617d1dbab844a2c9d72ab, which is a nice simplification.

File: src/init.cpp [hunk 8]
```diff
@@ -205,7 +205,14 @@ void InitContext(NodeContext& node)
     g_shutdown.emplace();
 
     node.args = &gArgs;
-    node.shutdown = &*g_shutdown;
+    node.shutdown_signal = &*g_shutdown;
+    node.shutdown_request = [&node] {
+        assert(node.shutdown_signal);
+        if (!(*node.shutdown_signal)()) return false;
+        // Wake any threads that may be waiting for the tip to change.
```
Comment (theuni) [inline src/init.cpp:212]: To avoid a race, this needs to lock `kernel_notifications.m_tip_block_mutex` before calling notify. `(*node.shutdown_signal)()` doesn't necessarily need to happen under the lock, but a lock is required after the signal in order to synchronize. Otherwise the update could be missed and we could potentially wait forever.

This is currently an example of [this trap](https://www.modernescpp.com/index.php/c-core-guidelines-be-aware-of-the-traps-of-condition-variables/) (see the "An atomic predicate" example).

See another example (and solution) [here](https://developers.redhat.com/articles/2022/12/06/implementing-c20-atomic-waiting-libstdc#handling_the_various_type_and_platform_choices) (ctrl+f for "acquisition above is significant").

So this would become:
```c++
if (!(*node.shutdown_signal)()) return false;
// Wake any threads that may be waiting for the tip to change.
if (node.notifications) {
    {
        // synchronize the shutdown state
        LOCK(node.notifications->m_tip_block_mutex);
    }
    node.notifications->m_tip_block_cv.notify_all();
}
```

Comment (ryanofsky) [reply src/init.cpp:212]: Thanks, good catch. I knew that pthreads required the lock to be held while notifying to avoid the lost wakeup bug (["if predictable scheduling behavior is required"](https://linux.die.net/man/3/pthread_cond_broadcast)), and that c++ dropped this requirement, even recommending against it saying ["Doing so may be a pessimization..."](https://en.cppreference.com/w/cpp/thread/condition_variable/notify_all).

But I didn't know that c++ still requires the lock to be held *before* notifying, even though it now recommends against holding the lock *while* notifying.

It does make sense that after the notifying thread updates the shutdown state, some extra synchronization is needed to make sure the notification is not sent too early, before the waiting thread is actually waiting.

Suggested fix looks right to me. And it looks like this bug might have been present even before this commit, though was much less likely to happen since notify_all call was happening long after the shutdown flag was set, not immediately after.

Comment (theuni) [reply src/init.cpp:212]: Yeah, I think the "Doing so may be a pessimization..." language encourages this footgun, sadly.

See [here](https://github.com/isocpp/CppCoreGuidelines/issues/1272) for a discussion about inverting the language.

File: src/node/interfaces.cpp (hunk 7 above)
Comment (ryanofsky) [reply src/node/interfaces.cpp:947]: Either approach seems ok. I probably wouldn't have added the null check, but it does a provide a small convenience if miner is running while the node is restarted, since waitTipChanged will patiently wait for startup to complete and a new block to be connected, instead of returning 0 on the first waitTipChanged call and requiring the miner to make a second call.

Whether we keep this check or drop it, it would probably be good to update the `waitTipChanged` documentation for the method to say that it either returns 0 on startup (if null check is dropped) or waits for the first block to be connected on startup (if null check is kept).

Comment (ryanofsky) [reply src/node/interfaces.cpp:947]: re: https://github.com/bitcoin/bitcoin/pull/30967#discussion_r1778162781

> In terms of how the Template Provider works, it used to rely on this check, but now it just waits for `getTip()` to return a value before calling `waitTipChanged()`. .
> 
> When implemented as part of Bitcoin Core, as in #29432, it can't call `waitTipChanged()` earlier, because it has no way of knowing what to set for `current_tip`.

I may be misunderstanding, but this doesn't seem right to me. If a caller just wants to know when a new block is connected it shouldn't need to call  `getTip()` in a loop until it returns non-null before calling `waitTipChanged()`, if that is what this comment is saying. The code should just be able to call `waitTipChanged()` with the zero value, which will return as soon as node is started and it knows what the tip is. (But I do agree with Marco it might be better to make let waitTipChanged() take a std::optional hash and not treat 0 as special). If this doesn't work, it suggests there could be a bug and code is not setting `m_tip_block` reliably.

**Review by ryanofsky — APPROVED** [review]
Code review ACK fa73a1d45d91fe289554e1f9f572917af5ba9d91, adding fix for potential log wakeup bug and some new comments.

**Review by ryanofsky — APPROVED** [review]
Code review ACK fa22e5c430acaef9713d9a4b4b97bb3f4876f816 (just rebased since last review)

File: src/node/interfaces.cpp (hunk 7 above)
Comment (Sjors) [reply src/node/interfaces.cpp:947]: > The code should just be able to call `waitTipChanged()` with the zero value, which will return as soon as node is started and it knows what the tip is.

That indeed seems like a better approach, will change that for the TP.

**Comment by Sjors** [comment]
ACK fa22e5c430acaef9713d9a4b4b97bb3f4876f816

Manually tested `-stopatheight`, a fatal indexer error (by breaking some code), calling `bitcoin-cli stop`, manually quitting qt, calling `stop` form the GUI console.

**Review by Sjors — APPROVED** [review]
ACK fa22e5c430acaef9713d9a4b4b97bb3f4876f816

Manually tested `-stopatheight`, a fatal indexer error (by breaking some code), calling `bitcoin-cli stop`, manually quitting qt, calling `stop` form the GUI console.

**Review by sedited — APPROVED** [review]
ACK fa22e5c430acaef9713d9a4b4b97bb3f4876f816

Nice improvements to the shutdown handling since I last reviewed. The nits can be ignored.

**Comment by gburkhow** [comment]
contrib/guix/INSTALL.md maflcko:2409-less-boost
